Address overshooting using correct zone air temperature at previous timestep

lgu1234 opened the pull request:
When ThirdOrderBackwardDifference is used as a choice of ZoneAirHeatBalanceAlgorithm, ZTM1 is used to represent the zone temperature at previous time step. The variable is used in zone energy balance equation to predict system load and solve zone temperature at the current time step. Unfortunately, ZTM1 does not represent the value in the OnOff thermostat. The correct value is used as zone temperature at the previous time step in OnOff thermostat. Partially addresses #6546.

Pull request overview
---------------------
Please change this line to a description of the pull request, with useful supporting information including whether it is a new feature, or fixes a defect, a cross reference to any defects addressed in this PR, the type of changes to be made, and whether diffs are expected/justified based on this change.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [x] Code style (parentheses padding, variable names)
 - [x] Functional code review (it has to work!)
 - [x] If defect, results of running current develop vs this branch should exhibit the fix
 - [x] CI status: all green or justified
 - [x] Performance: CI Linux results include performance check -- verify this
 - [x] Unit Test(s)
 - C++ checks:
   - [x] Argument types
   - [x] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes n/a

lgu1234 committed:
Address overshooting using correct zone air temperature at previous time step

lgu1234 committed:
Merge branch 'develop' into 157643005-ddress-limitations-and-overshooting-problem-using-ZTM1

rraustad commented on src/EnergyPlus/ZoneTempPredictorCorrector.cc:
```diff
@@ -3608,7 +3608,10 @@ namespace ZoneTempPredictorCorrector {
+                    if (ZoneAirSolutionAlgo == Use3rdOrder) {
+                        Tprev = MAT(ZoneNum);
+                        if (ShortenTimeStepSys) Tprev = XMPT(ZoneNum);
+                    }
```
Instead of testing for shortened time step could ZTM1 be updated correctly so this test isn't needed here?

lgu1234 replied:
@rraustad You raised a good question. Here are several reasons I am reluctant to make changes of ZTM1 correctly in the original place.

1. ZTM1 is assigned by XMAT in line 3190 in ZoneTempPredictorCorrector
                ZTM1(ZoneNum) = XMAT(ZoneNum);
XMAT is assigned by ZTAV in line 4911 in ZoneTempPredictorCorrector
            XMAT(ZoneNum) = ZTAV(ZoneNum); // using average for whole zone time step.

2 When DownInterpolate4HistoryValues is called, the zone air temperature at previous time steps used in 3order equations are interpolated with lengths of previous time steps.

Therefore, ZTM1 is not exactly a value at previous time step.

The best way to fix is to remove downshifting, so that we don't have any issues for history terms. I proposed this idea before. 

My fix is to allow zone air temperature at previous time step is used correctly for this new feature only. If I change it in the originally place, it will cause many differences, so that it will not be a small fix.

mjwitte commented on src/EnergyPlus/ZoneTempPredictorCorrector.cc:
```diff
@@ -3608,7 +3608,10 @@ namespace ZoneTempPredictorCorrector {
                 if (TempCtrlFound && TempControlledZone(RelativeZoneNum).DeltaTCutSet > 0.0) {
```
@lgu1234 When this if is false, `Tprev` is uninitialized below at [line 3630](https://github.com/NREL/EnergyPlus/pull/6754/files#diff-26a720ee34299122c810df4c89b63101R3630).  Looks like the closing } in 3629 should be move down.  Similar problem appears in the SingleCoolingSetPoint block, but DualSetPoint looks OK.

lgu1234 replied:
@mjwitte You are right. It is a very good catch. Changes are done. Thanks.

mjwitte commented on src/EnergyPlus/ZoneTempPredictorCorrector.cc:
```diff
@@ -3608,7 +3608,10 @@ namespace ZoneTempPredictorCorrector {
+                        Tprev = MAT(ZoneNum);
+                        if (ShortenTimeStepSys) Tprev = XMPT(ZoneNum);
+                    }
                     if (ZoneAirSolutionAlgo == UseAnalyticalSolution || ZoneAirSolutionAlgo == UseEulerMethod) Tprev = ZoneT1(ZoneNum);
```
This should be an elseif or simply else. Same comment in other similar blocks.

lgu1234 replied:
@mjwitte I like it and am going to make changes.

mjwitte commented on src/EnergyPlus/ZoneTempPredictorCorrector.cc:
```diff
@@ -3608,7 +3608,10 @@ namespace ZoneTempPredictorCorrector {
+                        Tprev = MAT(ZoneNum);
+                        if (ShortenTimeStepSys) Tprev = XMPT(ZoneNum);
+                    }
                     if (ZoneAirSolutionAlgo == UseAnalyticalSolution || ZoneAirSolutionAlgo == UseEulerMethod) Tprev = ZoneT1(ZoneNum);
```
Also, all of this stuff related to onoff control should be skipped if there aren't any onoff thermostats in the idf.  As it stands now, this [loop to find RelativeZoneNumber](https://github.com/NREL/EnergyPlus/pull/6754/files#diff-26a720ee34299122c810df4c89b63101L3581) happens on every call, no matter what type of thermostat control these is or isn't. This whole RelativeZoneNum thing is unnecessary, but that's for another day. For now, mimic the staged thermostats and add `NumOnOffCtrZone` flag similar to `NumStageCtrZone ` that skips all of this if there aren't any.

Also, it seems that all of this (which was move here in #6499) would be better placed [here](https://github.com/NREL/EnergyPlus/blob/b60abaa44542b4495671647f18044141b63b1ac2/src/EnergyPlus/ZoneTempPredictorCorrector.cc#L3060) in `PredictSystemLoads` rather than `CalcPredictedSystemLoads`. This would make it similar to the staged thermostat and would put this before some of the zone air temp updates which might be a better fix than this one.

lgu1234 replied:
@mjwitte The section was moved into PredictSystemLoads.

lgu1234 committed:
Address comments

lgu1234 committed:
Merge branch 'develop' into 157643005-ddress-limitations-and-overshooting-problem-using-ZTM1

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 157643005-ddress-limitations-and-overshooting-problem-using-ZTM1

mjwitte commented:
@lgu1234 I do not see much difference in zone temperature behavior before and after this fix. Can you point me to which file/which times show improvement? Here are csv outputs for before, current, and the original fix on this branch.
Note that I merged in develop to bring this branch up to date.
[6546ToGu.zip](https://github.com/NREL/EnergyPlus/files/2117140/6546ToGu.zip)

lgu1234 commented:
@mjwitte Thanks for bringing the branch up to date. The previous upload addressed comments only:

1.	Initialize Tprev
2.	Use an elseif to cover two scenarios
3.	Move a section of code from CalcPredictedSystemLoad to PredictSystemLoads

Therefore, results are not changed. The fix was from the previous upload on May 18, 2018 to use correct zone air temperature at previous time step, instead of using ZTM1 when ZoneAirSolutionAlgo = Use3rdOrder.

mjwitte commented:
@lgu1234 Which test file shows that this branch makes a change in results compared to current develop?

lgu1234 commented:
@mjwitte You may try a test file in https://github.com/NREL/EnergyPlusDevSupport/blob/master/DefectFiles/6000s/6546/6546Detailed.idf.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 157643005-ddress-limitations-and-overshooting-problem-using-ZTM1

mjwitte commented on src/EnergyPlus/ZoneTempPredictorCorrector.cc:
```diff
@@ -3416,6 +3421,147 @@ namespace ZoneTempPredictorCorrector {
 
+        // Setpoint revision for onoff thermostat
+        Real64 TempTole = 0.02;
+        Real64 Tprev;
```
These two lines can move inside the main if block below.  No need for these to be created if there aren't any NumOnOffCtrZone.

lgu1234 replied:
@mjwitte Yes. I move these two line inside the if block.

mjwitte commented on src/EnergyPlus/ZoneTempPredictorCorrector.cc:
```diff
@@ -3416,6 +3421,147 @@ namespace ZoneTempPredictorCorrector {
+                        }
+                        else {
+                            Tprev = ZoneT1(ZoneNum);
+                        }
```
This section that sets `Tprev` looks identical for all three thermostat types, so it could appear just once before line 3442 unless I'm missing something.

mjwitte replied:
This comment refers to lines 3445:3451

lgu1234 replied:
@mjwitte I agree. I am going to make changes.

mjwitte reviewed (approved):
@lgu1234 This is almost ready to go in. This gives a nice little performance boost by avoiding some code when there aren't any on/off tstats. Just a couple more minor changes and one more round of update to current develop should finish this.

See more comments over in #6546.

lgu1234 committed:
Address comments

lgu1234 committed:
Merge branch '157643005-ddress-limitations-and-overshooting-problem-using-ZTM1' of https://github.com/NREL/EnergyPlus into 157643005-ddress-limitations-and-overshooting-problem-using-ZTM1

lgu1234 committed:
Merge branch 'develop' into 157643005-ddress-limitations-and-overshooting-problem-using-ZTM1
